### DISCUSSION TITLE
Upscaling methods

### DISCUSSION THREAD

**Discussion by Novicek** [ihhub/fheroes2#5016] [Ideas]
Hello again,

This is a question/request about whether this project will receive alternative upscaling algorithms like Dosbox or Verok's mod do.

Currently I much prefer playing the Windows version with the mod as the provided filters make the game look way prettier (unfortunately, it does not allow playing in higher resolutions like this project does, nor any of the other myriad of improvements.)

**Comment by zenseii** [comment]
Hi,
There are frequently discussions about this like this issue thread:
https://github.com/ihhub/fheroes2/issues/4946

I believe there are other places it has been discussed too. Currently there is no concrete work being done to implement this but it might in the future.

**Comment by a1exsh** [comment]
Upscaling!..  But of course!

![image](https://user-images.githubusercontent.com/489601/209409301-6a16b8df-066e-4857-8ec0-1d151916aac6.png)

Can you see what I did here? :)

I've noticed that there is sort of a recurring complaint from the less versed players that the game graphics is "not good/modern".  It may mean many things, but I can guess that part of it is justified given that the original game resources are of too low resolution for the contemporary Full HD displays.

What if we look at it from a different angle: not an advanced screen filter, but rather advanced scaling algorithm?  I've taken a free to use [Upscaler](https://gitlab.com/TheEvilSkeleton/Upscaler) and ran the Bone Dragon sprite through it.  I did expect some improvement over our default linear scaling, but the results have simply astonished me (and my kids, as you can guess). :D

![border_700x525_evil_broken](https://user-images.githubusercontent.com/489601/209409886-b6a8018b-b2f1-4ec2-821f-52ef37d5b0c3.png)


Now, of course, it would be impractical to do this sort of advanced upscaling on the fly, or to distribute the upscaled resources (still subject to copyright), but what if we provide a tool for the users to automatically unpack the resources from AGG, convert to PNG and then run upscaling each individual image?..

That sounds feasible to me, as it needs to be done just once, and then once can enjoy crisp Full HD consistent with the original artistic style.

**Comment by Novicek** [comment]
Honestly, with the 1.0.0 release, I'm almost happy with how the game looks on PC when I set it to software fullscreen renderer and nearest neighbor upscaling. If the game also had integer scaling option (where it would resize only to the maximum multiple of your selected resolution that fits into your screen resolution and fill the rest with black bars), we'd at least have accurate graphics without artifacting.

I suppose I could achieve it already if I played it with hardware upscaling and set up integer scaling in nvidia, but I avoid that for multiple reasons (can't set it on a per-game basis and I don't want it for other games and it will also go back to resizing my desktop and slow alt-tabbing).

Edit: Actually I was able to get pixel perfect upscaling. Since I am using a 3840x2160 resolution (4K), I just had to find the proper resolution that would scale to it well, which is 720x540. However, as it was not an option in-game and editing custom resolution in the config also didn't help, I went to the nvidia settings and added it as a new custom resolution. Then it showed up in game as a selectable option and it works quite well :)

Edit 2: Here's how it looks upscaled to 4K. Notice the small font. On a non multiple resolution, the thickness of it would be uneven in places.
<img width="1920" alt="Untitled" src="https://user-images.githubusercontent.com/17748367/210035518-1fd4c27d-0c3e-40b4-a106-e4a8009159e1.png">

**Reply by a1exsh** [reply]
> Honestly, with the 1.0.0 release, I'm almost happy with how the game looks on PC when I set it to software fullscreen renderer and nearest neighbor upscaling. If the game also had integer scaling option (where it would resize only to the maximum multiple of your selected resolution that fits into your screen resolution and fill the rest with black bars), we'd at least have accurate graphics without artifacting.

I've started a proof-of-concept work where I'm trying to achieve exactly that, as a first step.  If that turns out to be a working approach, I'm aiming to plug in other upscaling algorithms, such as xBRz.  For example, compare the effect of 4x resizing/upscaling on that same Bone Dragon sprite (original, xBRz, ESRGAN):

![image](https://user-images.githubusercontent.com/489601/210048685-ca9e724b-2ca4-4416-9e88-fe302080abf6.png) ![image](https://user-images.githubusercontent.com/489601/210048702-15ec2f19-19b4-49aa-acce-5944b197def0.png)
![image](https://user-images.githubusercontent.com/489601/210048728-de8c0232-3108-46fb-93cf-5c558caaff8a.png)

The latest variant shows a great deal more details, but even xBRz is quite an improvement over the nearest neigbhour scaling.  Where it can really shine, however, is Castle screen.  Just look at these two images to feel the difference (2x upscaling):

![Necromancer Castle - Resized - 1280x496](https://user-images.githubusercontent.com/489601/210049003-9a4184a4-aae2-4f82-a42e-1fcfbebfe466.png)

![Necromancer Castle - Upscaled - 1280x496](https://user-images.githubusercontent.com/489601/210049004-a84e68f0-e494-4cc9-ac13-5687e902b783.jpg)

> I suppose I could achieve it already if I played it with hardware upscaling and set up integer scaling in nvidia, but I avoid that for multiple reasons (can't set it on a per-game basis and I don't want it for other games and it will also go back to resizing my desktop and slow alt-tabbing).
> 
> Edit: Actually I was able to get pixel perfect upscaling. Since I am using a 3840x2160 resolution (4K), I just had to find the proper resolution that would scale to it well, which is 720x540. However, as it was not an option in-game and editing custom resolution in the config also didn't help, I went to the nvidia settings and added it as a new custom resolution. Then it showed up in game as a selectable option and it works quite well :)
> 
> Edit 2: Here's how it looks upscaled to 4K. Notice the small font. On a non multiple resolution, the thickness of it would be uneven in places. <img alt="Untitled" width="1920" src="https://user-images.githubusercontent.com/17748367/210035518-1fd4c27d-0c3e-40b4-a106-e4a8009159e1.png">

Yes, upscaling by a whole multiple is critical for the font rendering, and this is what I'm working on.  Hope to be able to release a working prototype soon. :)

**Reply by ihhub** [reply]
Choosing an algorithm for up scaling is not an issue at all. Our codebase already has everything for this. Before implementing any of these algorithms please measure the timing and performance drop. If it's a matter of 10-20% then it's fine but if an algorithm makes rendering in 3-5 times slower as they usually do then the answer is a no for the implementation, unless someone optimize it properly.

**Reply by a1exsh** [reply]
> Choosing an algorithm for up scaling is not an issue at all. Our codebase already has everything for this. Before implementing any of these algorithms please measure the timing and performance drop. If it's a matter of 10-20% then it's fine but if an algorithm makes rendering in 3-5 times slower as they usually do then the answer is a no for the implementation, unless someone optimize it properly.

@ihhub good points! :)  To open up the cards a little, the approach I've taken is scaling the resources only once at load time, and not constantly during rendering, so in general it is less sensitive to the potential performance impact.  I have also not noticed any impact on the battle screen while running in Full HD and 2x pre-resized sprites, but I will also try to measure it as well.

Edit: of course, pre-scaling at load time is not an option for ESRGAN, as it will take forever.  As I've mentioned initially, for this to work, the idea is to convert resources offline and then load from alternative path (not from the original AGG files).  I haven't measured it, but I'm optimistic that xBRz can work just fine with resizing at load time.

**Reply by ihhub** [reply]
The engine can't handle resized assets unless we rewrite it. Plus don't forget about transform layer of every image / sprite. All special effects on images are done through this layer.

**Reply by a1exsh** [reply]
> The engine can't handle resized assets unless we rewrite it. Plus don't forget about transform layer of every image / sprite. All special effects on images are done through this layer.

Yes, in its current form it cannot work, but I've found a way to make a focused change that only updates the engine, and keeps most of the UI code untouched, so all the hard-coded measurements continue to work just fine. ;)

**Reply by ihhub** [reply]
screen.cpp file is designed to be single-threaded and 8-bit to 32-bit conversion eats a huge chunk of resources. This is a very simple operation with complexity of O(n) where n is the size of an image in pixels. I assume that the complexity of any upscaling algorithm with multiple pixel interactions is > O(n * m) where m is the number of pixels used for a single iteration of the algorithm. In this case single-threaded code would be much slower even on average machines.

**Reply by Laserlicht** [reply]
@a1exsh Really good work! Please follow this up. 🙂

I did not know ESRGAN until now. But the town screen looks really cool. Almost like a new game.

Would be cool if we find a solution. The option to have such filters would be really great.

I think having that calculated once would not be a problem. Then it may well take something when you initiate the process.

**Comment by a1exsh** [comment]
Here is the version 1.0.0 running in 1920x1080 resolution:
https://user-images.githubusercontent.com/489601/210100333-5808f403-060a-4e82-8d22-9e353c754afd.mp4

And here is my locally built version with the 2x2 upscaling at the same resolution for comparison:
https://user-images.githubusercontent.com/489601/210100395-d8b017b0-032b-45b0-9c90-12998bdd3570.mp4

I do not notice any performance difference when working with upscaled resources.  The FPS is lower than normal for my machine in both cases due to recording.

Note the upscaled cursor and bigger menu / buttons that make it easier to use — this is especially helpful on even higher resolutions. ;)

**Reply by ihhub** [reply]
Hi @a1exsh , set hero speed on Adventure Map to 9 and walk him over the map. You should have 125 FPS. This is the correct measurement of speed. Most of players complain about speed only on Adventure Map.

**Reply by a1exsh** [reply]
> Hi @a1exsh , set hero speed on Adventure Map to 9 and walk him over the map. You should have 125 FPS. This is the correct measurement of speed. Most of players complain about speed only on Adventure Map.

@ihhub thank you for the suggestion — I had to fix a nasty rendering issue first, but now I've tried it and I see even a slight improvement in my version over 1:1 rendering on 1920x1080 resolution.  The unpatched code gives me slightly above 100 FPS, and the patched one — 110-120 FPS.  I guess this is because there is roughly four times less calls to render individual tile on the adventure map, while each tile is two times bigger in each dimension.

I hope to be able cleanup the code so that I can open a PR today. :)

**Reply by ihhub** [reply]
How much FPS do you have without your changes? 125 FPS?

**Reply by a1exsh** [reply]
> How much FPS do you have without your changes? 125 FPS?

As I said, unpatched — 100-110 FPS, patched — 110-120 (yes, more!)

**Reply by ihhub** [reply]
So it means that you reduced the effective resolution by a factor of 2 for each axis which should be roughly in 3.5 times faster than you added the algorithm which makes the performance in about 3 times slower. The result is slightly faster but the number of objects is smaller. Is it correct?

**Reply by a1exsh** [reply]
> So it means that you reduced the effective resolution by a factor of 2 for each axis which should be roughly in 3.5 times faster than you added the algorithm which makes the performance in about 3 times slower. The result is slightly faster but the number of objects is smaller. Is it correct?

No, not exactly, as actual rendering on the display is not using any new algorithm.  Assuming most of the time for rendering a single adventure map frame is spent `Blit`-ing the tiles, we are now `Blit`-ing ~4 times _less_ tiles in total per frame, but each tile is exactly four times _bigger_ than it used to be.  No scaling is happening at this time, as all tiles are pre-scaled while being loaded.  All in all, we are covering the same screen area, but in "rougher strokes". ;)
